File: apps/tauri-app/src/lib/utils/undoOffer.ts
// The way back from the last History action: what the strip says, when it
// stops offering, and what is remembered of an action that stopped on a
// conflict. Pure, and mirrored natively by `Services/UndoOffer.swift`.

import type { OperationInProgress, UndoPoint, UndoStart } from '$lib/api/commands'

/** A History action that can be taken back, with what its sentence names. */
export type UndoableAction =
  { kind: 'cherryPick'; target: string } | { kind: 'squash' } | { kind: 'reorder' }

/**
 * An Undo on offer. It lives in client memory only: after a restart the
 * branch's reflog is the way back.
 */
export interface UndoOffer {
  /** The repository the action ran in — never whichever one is open by now. */
  repoPath: string
  /** What the action did, as the strip states it. */
  sentence: string
  /** What core needs to put the branch back. */
  point: UndoPoint
  /**
   * The commits the action was asked for, newest first: what History selects
   * again once they are back.
   */
  restores: string[]
}

/**
 * What the strip says an action did. `count` is the number of commits it was
 * asked for — what an Undo brings back, whatever a Continue made of them.
 */
export function landedSentence(action: UndoableAction, count: number): string {
  const commits = count === 1 ? '1 commit' : `${count} commits`
  switch (action.kind) {
    case 'cherryPick':
      return `Cherry-picked ${commits} onto “${action.target}”.`
    case 'squash':
      return `Squashed ${commits} into one.`
    case 'reorder':
      return `Reordered ${commits}.`
  }
}

/**
 * Whether the offer still stands under `status`. It falls once the status shows
 * the branch checked out at any commit but the one the action left it on — a
 * commit, an amend, a pull or another rewrite has moved it, here or in a
 * terminal — because an Undo that can only be refused is not left on screen to
 * fail. While another branch is checked out, or HEAD is detached, the status
 * cannot see that branch's tip: the offer stays, and core is the judge.
 */
export function undoStillStands(
  point: UndoPoint,
  status: { branch: string; detached: boolean; headSha: string }
): boolean {
  if (status.detached || status.branch !== point.branch) return true
  return status.headSha === point.after_sha
}

/**
 * A History action that stopped on a conflict and left its operation open:
 * what is kept until the operation ends, so that a Continue which lands can
 * still be taken back and an Abort knows which branch to return to. Written
 * only by the app's own action, so an operation begun in a terminal has none.
 */
export interface OpenAction {
  /** The repository the action ran in. */
  repoPath: string
  /** What the action left open: a cherry-pick, or a squash's or reorder's rebase. */
  operation: OperationInProgress
  /** Where the action began, from core. */
  start: UndoStart
  action: UndoableAction
  /** The commits the action was asked for, newest first. */
  asked: string[]
}

/**
 * Whether `open` is still the operation `status` shows. It is forgotten the
 * moment the status shows anything else — aborted, finished or quit, here or
 * in a terminal. A cherry-pick is also held to its branch; a rebase cannot be,
 * because it detaches HEAD and the status names no branch while one is open.
 */
export function openActionStillStands(
  open: OpenAction,
  status: { operation: OperationInProgress | null; branch: string }
): boolean {
  if (status.operation !== open.operation) return false
  return open.operation === 'Rebase' || status.branch === open.start.branch
}

/**
 * The undo point for `open`, now that a Continue has carried its operation to
 * the end and `status` has been read again — or null when none can be vouched
 * for:
 *
 * - the status does not show the action's branch checked out, so its new tip
 *   cannot be read;
 * - `startedAt`, where git's own state says the operation began, is not where
 *   this action began — the open operation was not the one it left — or could
 *   not be read at all, which vouches for nothing;
 * - the branch is where it began, so there is nothing to take back.
 *
 * A wrong `after_sha` would be harmless — core refuses a tip that is not the
 * branch's — but nothing in core relates `before_sha` to it, which is why the
 * checks here are about the start.
 */
export function undoPointAfterContinue(
  open: OpenAction,
  startedAt: string | null,
  status: { branch: string; detached: boolean; headSha: string }
): UndoPoint | null {
  const { start } = open
  if (status.detached || status.branch !== start.branch || !status.headSha) return null
  if (startedAt !== start.before_sha) return null
  if (status.headSha === start.before_sha) return null
  return { ...start, after_sha: status.headSha }
}

File: apps/swift-ui-app/Sources/LeoGit/Services/UndoOffer.swift
import Foundation

/// The way back from the last History action: what the strip says, and when it
/// stops offering. It lives in client memory only — after a restart the
/// branch's reflog is the way back. Mirrors the Tauri client's `undoOffer.ts`,
/// as `OpenAction` below does.
struct UndoOffer: Equatable {
    /// The History actions that can be taken back.
    enum Action: Equatable {
        /// Onto the named branch.
        case cherryPick(target: String)
        case squash
        case reorder
    }

    /// What the action did, as the strip states it.
    let sentence: String
    /// What core needs to put the branch back.
    let point: UndoPoint
    /// The commits the action was asked for, newest first: what History
    /// selects again once they are back.
    let restores: [String]

    /// The offer an action leaves behind. `asked` are the commits it was asked
    /// for, newest first — what an Undo brings back, whatever a Continue made
    /// of them.
    init(after action: Action, of asked: [String], point: UndoPoint) {
        sentence = Self.landedSentence(action, count: asked.count)
        self.point = point
        restores = asked
    }

    /// What the strip says an action did to `count` commits.
    static func landedSentence(_ action: Action, count: Int) -> String {
        let commits = count == 1 ? "1 commit" : "\(count) commits"
        return switch action {
        case let .cherryPick(target): "Cherry-picked \(commits) onto “\(target)”."
        case .squash: "Squashed \(commits) into one."
        case .reorder: "Reordered \(commits)."
        }
    }

    /// Whether the offer still stands under `status`. It falls once the status
    /// shows the branch checked out at any commit but the one the action left
    /// it on — a commit, an amend, a pull or another rewrite has moved it, here
    /// or in a terminal — because an Undo that can only be refused is not left
    /// on screen to fail. While another branch is checked out, or HEAD is
    /// detached, the status cannot see that branch's tip: the offer stays, and
    /// core is the judge.
    func stillStands(under status: RepoStatus) -> Bool {
        if status.detached || status.branch != point.branch { return true }
        return status.headSha == point.afterSha
    }
}

/// A History action that stopped on a conflict and left its operation open:
/// what is kept until the operation ends, so that a Continue which lands can
/// still be taken back and an Abort knows which branch to return to. Git keeps
/// no record of either. Written only by the app's own action, so an operation
/// begun in a terminal has none.
struct OpenAction: Equatable {
    /// What the action left open: a cherry-pick, or a squash's or reorder's
    /// rebase.
    let operation: OperationInProgress
    /// Where the action began, from core.
    let start: UndoStart
    let action: UndoOffer.Action
    /// The commits the action was asked for, newest first.
    let asked: [String]

    /// Whether this is still the operation `status` shows. It is forgotten the
    /// moment the status shows anything else — aborted, finished or quit, here
    /// or in a terminal, and a status that is not there at all. A cherry-pick
    /// is also held to its branch; a rebase cannot be, because it detaches HEAD
    /// and the status names no branch while one is open.
    func stillStands(under status: RepoStatus?) -> Bool {
        guard let status, status.operation == operation else { return false }
        return operation == .rebase || status.branch == start.branch
    }

    /// The undo point, now that a Continue has carried the operation to its end
    /// and `status` has been read again — or `nil` when none can be vouched
    /// for:
    ///
    /// - the status does not show the action's branch checked out, so its new
    ///   tip cannot be read;
    /// - `startedAt`, where git's own state says the operation began, is not
    ///   where this action began — the open operation was not the one it left
    ///   — or could not be read at all, which vouches for nothing;
    /// - the branch is where it began, so there is nothing to take back.
    ///
    /// A wrong `afterSha` would be harmless — core refuses a tip that is not
    /// the branch's — but nothing in core relates `beforeSha` to it, which is
    /// why the checks here are about the start.
    func undoPoint(startedAt: String?, under status: RepoStatus) -> UndoPoint? {
        guard !status.detached, status.branch == start.branch, !status.headSha.isEmpty else {
            return nil
        }
        guard startedAt == start.beforeSha else { return nil }
        guard status.headSha != start.beforeSha else { return nil }
        return UndoPoint(
            branch: start.branch,
            beforeSha: start.beforeSha,
            afterSha: status.headSha,
            returnBranch: start.returnBranch
        )
    }
}
